File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.cpp
class Solution {
public:
    struct Node{
        int first;
        int second;
        int distance;

        Node( int first, int second,int distance): first(first) , second(second), distance(distance){};

    };




    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();
        queue<Node> q;
        vector<vector<int>> visited(n, vector<int>(m,0));

        if(grid[0][0] == 1 || grid[n - 1][m-1] == 1) return -1;
        Node node(0,0,1);
        visited[0][0] = 1;
        q.push(node);

        while(!q.empty())
        {
            Node curr = q.front();
            q.pop();

            if(curr.first == n- 1 && curr.second == m - 1 ) return curr.distance;

            int dirs [8][2]={{1,0},{0,1},{-1,0},{0,-1},{1,1},{-1,-1},{-1,1},{1,-1}};

            for(auto dir : dirs)
            {
                int newRow = curr.first + dir[0];
                int newCol = curr.second + dir[1];

                if(newRow < n && newRow >=0 && newCol < m && newCol >=0 && visited[newRow][newCol] == 0 && grid[newRow][newCol] == 0)
                {
                    visited[newRow][newCol] = 1;
                    Node temp(newRow, newCol , curr.distance + 1 );
                    q.push(temp);
                }
            }

            





        }
        return -1;
    }
};

File: 1171-shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.java
class Solution {

    class Pair {
        int r, c, dist;

        Pair(int r, int c, int dist){
            this.r = r;
            this.c = c;
            this.dist = dist;
        }
    }

    public int shortestPathBinaryMatrix(int[][] grid) {

        int n = grid.length;

        // start or end blocked
        if(grid[0][0] == 1 || grid[n-1][n-1] == 1)
            return -1;

        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(0,0,1));

        // mark visited
        grid[0][0] = 1;

        int[][] dirs = {
                {0,1},{1,0},{1,1},
                {-1,-1},{0,-1},{-1,0},
                {-1,1},{1,-1}
        };

        while(!q.isEmpty()){

            Pair curr = q.poll();

            if(curr.r == n-1 && curr.c == n-1)
                return curr.dist;

            for(int[] d : dirs){

                int nr = curr.r + d[0];
                int nc = curr.c + d[1];

                if(nr>=0 && nc>=0 && nr<n && nc<n && grid[nr][nc]==0){

                    q.offer(new Pair(nr, nc, curr.dist+1));
                    grid[nr][nc] = 1; // mark visited
                }
            }
        }

        return -1;
    }
}
